docs/cl/libvirt.md

```diff
@@ -0,0 +1,245 @@
+
+You will need the `virsh` and `qemu-img` binaries installed on your system:
+
+```
+sudo dnf install libvirt-client qemu-img
+```
+
+## CoreOS Base image
+
+You will need to manually download a CoreOS base image. Qemu supports a copy-on-write format, so this will allow for deduplication.
+
+```
+wget https://stable.release.core-os.net/amd64-usr/current/coreos_production_qemu_image.img.bz2
+bunzip2 coreos_production_qemu_image.img.bz2
+qemu-img resize coreos_production_qemu_image.img +8G
+```
+
+Make a note of the absolute path to this image, you'll need it later.
+
+
+## DNS
+
+Libvirt will start a DNS server for each cluster, and will create a DNS name for every node. However, we need to make the autogenerated names resolvable on the local machine. In order to do this, you will need to put NetworkManager in to `dnsmasq` mode, which will route all DNS queries through a local dnsmasq instance. Then we can instruct dnsmasq to delegate queries for a specific domain to the libvirt resolver.
+
+You will need to select values for your `domain` and, if overridden, `node_ip_pool` Terraform variables below. Use them here.
+
+For example, if you use a `domain` of `hestia.k8s` and the default  `node_ip_pool` of `192.168.120.0/24`, then your DNS server will be at `192.168.120.1`
+
+1. Edit `/etc/NetworkManager/NetworkManager.conf` and set `dns=dnsmasq` in section `[main]`
+2. Tell dnsmasq to use your cluster. The syntax is `server=/<domain>/<firstIP>`. For this example:
+```
+echo server=/hestia.k8s/192.168.120.1 | sudo tee /etc/NetworkManager/dnsmasq.d/typhoon.conf
+```
+3. `systemctl restart NetworkManager`
+
+## Terraform Setup
+
+Install [Terraform](https://www.terraform.io/downloads.html) v0.11.x on your system.
+
+```sh
+$ terraform version
+Terraform v0.11.1
+```
+
+Install [terraform-provider-libvirt ](https://github.com/dmacvicar/terraform-provider-libvirt).
+
+```sh
+ go get github.com/dmacvicar/terraform-provider-libvirt
+mkdir -p ~/.terraform.d/plugins
+cp $GOPATH/bin/terraform-provider-libvirt ~/.terraform.d/plugins/
+```
+
+Install [terraform-provider-ct ](https://github.com/coreos/terraform-provider-ct).
+```sh
+ go get github.com/coreos/terraform-provider-ct
+mkdir -p ~/.terraform.d/plugins
+cp $GOPATH/bin/terraform-provider-ct ~/.terraform.d/plugins/
+```
+
+Read [concepts](../architecture/concepts.md) to learn about Terraform, modules, and organizing resources. Change to your infrastructure repository (e.g. `infra`).
+
+```
+cd infra/clusters
+```
+
+## Cluster
+
+Define a Kubernetes cluster using the module `bare-metal/container-linux/kubernetes`.
+
+```tf
+module "libvirt-hestia" {
+  source = "git::https://github.com/poseidon/typhoon//libvirt/container-linux/kubernetes?ref=v1.11.0"
+  
+  cluster_name    = "hestia"
+  base_image_path = "/home/user/coreos.img"
+  domain          = "hestia.k8s"
+
+  controller_names = ["node1", "node2"]
+  controller_ips   = ["192.168.120.10", "192.168.120.11"]
```

Comment:
No IPs. Users should not be permitted to hardcode IPs.

Reply:
Having stable IPs for the controllers makes some things easier for DNS. They could be autogenerated from the `node_ip_pool`. The only downside to this is that they could change if the controller count changes.